owsi-core/owsi-core-components/owsi-core-component-jpa/src/main/java/fr/openwide/core/jpa/batch/executor/MultithreadedBatch.java

```diff
@@ -0,0 +1,132 @@
+package fr.openwide.core.jpa.batch.executor;
+
+import java.math.BigDecimal;
+import java.util.Date;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.config.ConfigurableBeanFactory;
+import org.springframework.context.annotation.Scope;
+import org.springframework.stereotype.Component;
+import org.springframework.transaction.TransactionStatus;
+import org.springframework.transaction.support.TransactionCallback;
+
+import com.google.common.collect.Lists;
+
+import fr.openwide.core.jpa.batch.processor.ThreadedProcessor;
+import fr.openwide.core.jpa.batch.util.ProcessorProgressLogger;
+import fr.openwide.core.spring.property.service.IPropertyService;
+
+@Component
+@Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+public class MultithreadedBatch extends AbstractBatch<MultithreadedBatch> {
+	
+	private static final Logger LOGGER = LoggerFactory.getLogger(MultithreadedBatch.class);
+	
+	private static final Logger PROGRESS_LOGGER = LoggerFactory.getLogger(ProcessorProgressLogger.class);
+	
+	@Autowired
+	private IPropertyService propertyService;
+	
+	private int threads = 2;
+	
+	private int timeoutInMinutes = 15;
+	
+	public MultithreadedBatch threads(int threads) {
+		this.threads = threads;
+		return this;
+	}
+	
+	public MultithreadedBatch timeoutInMinutes(int timeout) {
+		this.timeoutInMinutes = timeout;
+		return this;
+	}
+	
+	public void run(String context, final List<Long> entityIds, final BatchRunnable<Long> batchRunnable) {
+		Date startTime = new Date();
+		
+		LOGGER.info("Beginning batch for %1$s: %2$d objects", context, entityIds.size());
+		
+		LOGGER.info("    preExecute start");
+		
+		writeTransactionTemplate.execute(new TransactionCallback<Void>() {
```

Comment:
You mean, apply Spring's proxies on the runnable before executing it? Seems like an attractive solution, indeed.

However... It doesn't change the fact that the read-only-or-read-write state of the runnable would be implicit by default. I mean, if one executes a runnable, nothing in his code will scream to his ears "this is read-only" or "this is read/write". So we'll end up in one of the following cases:
 * If the default is read/write, there might be unwanted side-effects to the executions
 * If the default is read-only, there might be some frustration from the users' perspective ("why the hell did this not change anything in database?")

This is an issue we don't have currently with name-based read-only detection (generally when we perform read-only work, we do in in a method named `list*`, `get*` or `search*`).

One solution would be to have two abstract `IBatchRunnable`s:
 * `ReadOnlyBatchRunnable`, with `@Transactionnal(readonly = true)" on every method
 * `ReadWriteBatchRunnable`, with `@Transactionnal(readonly = false)" on every method
This would indeed provide a way to customize exactly which method is readonly and which is not, while making obvious in the user's code whether something is read-only or not.

The other solution I suggested above (having two methods on the executor, `runReadOnly` and `runReadWrite`) would indeed be less customizable.

I'd be glad to discuss this further on mattermost.